src/main/java/graphql/annotations/dataFetchers/MethodDataFetcher.java

```diff
@@ -118,40 +132,67 @@ private Object[] invocationArgs(DataFetchingEnvironment environment, ProcessingE
             Parameter[] parameters = constructor.getParameters();
-            if (parameters.length == 1 && parameters[0].getType().isAssignableFrom(arg.getClass())) {
-                return constructNewInstance(constructor, arg);
+
+            if (parameters.length == 1 && optionalArg.isPresent() && parameters[0].getType().isAssignableFrom(optionalArg.get().getClass())) {
+                if (parameters[0].getType().isAssignableFrom(Optional.class)) {
+                    return constructNewInstance(constructor, arg);
+                } else {
+                    return constructNewInstance(constructor, optionalArg.orElse(null));
+                }
             } else {
                 List<Object> objects = new ArrayList<>();
-                Map map = (Map) arg;
+                Map map = (Map) optionalArg.orElseGet(Collections::emptyMap);
                 for (Parameter parameter : parameters) {
                     String name = toGraphqlName(parameter.getAnnotation(GraphQLName.class) != null ? parameter.getAnnotation(GraphQLName.class).value() : parameter.getName());
-                    objects.add(buildArg(parameter.getParameterizedType(), ((GraphQLInputObjectType) graphQLType).getField(name).getType(), map.get(name)));
+                    if (!map.containsKey(name)) {
+                        objects.add(null);
+                    } else {
+                        objects.add(buildArg(parameter.getParameterizedType(), ((GraphQLInputObjectType) graphQLType).getField(name).getType(), map.get(name)));
+                    }
                 }
                 return constructNewInstance(constructor, objects.toArray(new Object[objects.size()]));
             }
         } else if (p instanceof ParameterizedType && graphQLType instanceof GraphQLList) {
-            List<Object> list = new ArrayList<>();
-            Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
-            GraphQLType wrappedType = ((GraphQLList) graphQLType).getWrappedType();
+            if (((ParameterizedType) p).getRawType() == Optional.class) {
+                if (!optionalArg.isPresent()) {
+                    return null;
+                } else {
+                    Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
+                    return Optional.ofNullable(buildArg(subType, graphQLType, arg));
+                }
+            } else {
+                List<Object> list = new ArrayList<>();
+                Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
+                GraphQLType wrappedType = ((GraphQLList) graphQLType).getWrappedType();
 
-            for (Object item : ((List) arg)) {
-                list.add(buildArg(subType, wrappedType, item));
+                for (Object item : ((List) optionalArg.orElseGet(Collections::emptyList))) {
+                    list.add(buildArg(subType, wrappedType, item));
+                }
+                return list;
+            }
```

Comment:
Keep this whole case unchanged, and move the next case `(p instanceof ParameterizedType && ((ParameterizedType) p).getRawType() == Optional.class)` before the list one